### PR TITLE
Encode key with slot id

### PR DESCRIPTION
**Pull request by shangxiaoxiong**
Encode key with slot id, so it will be efficient to migrate keys by slot when scale

Also support CLUSTER KEYSLOT command

### CODE REVIEW

File: kvrocks.conf [hunk 1]
```diff
@@ -200,6 +200,12 @@ max-backup-to-keep 1
 # default: 1 day
 max-backup-keep-hours 24
 
+# Enable the kvrocks to support cluster mode that encoding key with slot id, if the db enabled
```
Comment (ShooterIT) [inline kvrocks.conf]: these lines are too long

File: src/redis_db.cc [hunk 2]
```diff
@@ -355,7 +433,8 @@ rocksdb::Status Database::Type(const Slice &user_key, RedisType *type) {
 }
 
 void Database::AppendNamespacePrefix(const Slice &user_key, std::string *output) {
-  ComposeNamespaceKey(namespace_, user_key, output);
+
```
Comment (ShooterIT) [inline src/redis_db.cc]:
```suggestion
```
remove empty line

File: tests/tcl/tests/unit/cluster.tcl [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+    test {CLUSTER KEYSLOT} {
+
+        set slot_table_len [llength $::CRC16_SLOT_TABLE]
+
+        for {set i 0} {$i < $slot_table_len} {incr i} {
+            assert_equal $i [r cluster keyslot [lindex $::CRC16_SLOT_TABLE $i]]
+        }
+    }
+}
```
Comment (ShooterIT) [inline tests/tcl/tests/unit/cluster.tcl]:
```suggestion
}

```
add one more empty line

File: src/version.h [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#pragma once
```
Comment (ShooterIT) [inline src/version.h]: remove this file

File: src/redis_db.cc [hunk 5]
```diff
@@ -206,17 +240,60 @@ rocksdb::Status Database::Scan(const std::string &cursor,
...
+    }
+
+    if (++slot_id >= HASH_SLOTS_SIZE) {
+      break;
+    }
+
+    if (slot_id > slot_start + HASH_SLOTS_MAX_ITERATIONS) {
+      if (keys->empty()) {
+        if (iter->Valid()) {
+
```
Comment (ShooterIT) [inline src/redis_db.cc]:
```suggestion
```

File: kvrocks.conf [hunk 6]
```diff
@@ -200,6 +200,12 @@ max-backup-to-keep 1
...
+# the cluster mode at first open, this option must not be disabled after restarted, and vice versa
+#
+# Defalut: yes
+cluster-enabled yes
```
Comment (ShooterIT) [inline kvrocks.conf]: by default, it should be `no`
```suggestion
# Defalut: no
cluster-enabled no
```

File: src/redis_cmd.cc [hunk 7]
```diff
@@ -32,6 +32,8 @@
 
 namespace Redis {
 
+const std::string kCursorPrefix = "_";
```
Comment (ShooterIT) [inline src/redis_cmd.cc]: I think we should add comments for `scan` in doc/commands to say our cursor is not a muber, but 0 always is the end of cursor

**Comment by ShooterIT** [comment]
@shangxiaoxiong Maybe you should add a new rocksdb CF to identify cluster mode and check if it is safe to enable/disable cluster mode in config file. WDYT?

**Comment by shangxiaoxiong** [comment]
Good suggestion！l will do it right away. @ShooterIT

**Review by ShooterIT — DISMISSED** [review]

**Comment by ShooterIT** [comment]
We don't have `sortedint` type tcl tests, so currently some code is not tested, i hope you can supplement some basic tcl test cases.

**Comment by ShooterIT** [comment]
After discussion, i prefer to encode slot id into key **by default**, so it is easy to change from master-replica mode to cluster mode for new version, even encoding slot id doesn't support `namespace`.

File: src/redis_db.cc [hunk 8]
```diff
@@ -355,7 +433,7 @@ rocksdb::Status Database::Type(const Slice &user_key, RedisType *type) {
 }
 
 void Database::AppendNamespacePrefix(const Slice &user_key, std::string *output) {
-  ComposeNamespaceKey(namespace_, user_key, output);
+  ComposeNamespaceKey(namespace_, user_key, output, storage_->IsClusterEnabled());
```
Comment (git-hulk) [inline src/redis_db.cc]: How about moving the `is_cluster_enabled` as the first parameter that keeps the input params before outputs

File: src/redis_db.cc [hunk 9]
```diff
@@ -150,31 +159,43 @@ void Database::Keys(std::string prefix, std::vector<std::string> *keys, KeyNumSt
...
-      int32_t ttl = metadata.TTL();
-      stats->n_key++;
-      if (ttl != -1) {
-        stats->n_expires++;
-        if (ttl > 0) ttl_sum += ttl;
+
+  while (true) {
+    ns_prefix.empty() ? iter->SeekToFirst() : iter->Seek(ns_prefix);
+    for (; iter->Valid(); iter->Next()) {
+      if (!prefix.empty() && !iter->key().starts_with(ns_prefix)) {
```
Comment (git-hulk) [inline src/redis_db.cc]: Should be `ns_prefix`?

File: src/redis_db.cc [hunk 10]
```diff
@@ -150,31 +159,43 @@ void Database::Keys(std::string prefix, std::vector<std::string> *keys, KeyNumSt
...
+        keys->emplace_back(user_key);
       }
     }
-    if (keys) {
-      ExtractNamespaceKey(iter->key(), &ns, &user_key);
-      keys->emplace_back(user_key);
-    }
+
+    if (!storage_->IsClusterEnabled()) break;
+    if (prefix.empty()) break;
```
Comment (git-hulk) [inline src/redis_db.cc:191]: Why we need to break the loop when the prefix was empty(which means `keys *` pattern)

File: src/redis_db.cc [hunk 11]
```diff
@@ -150,31 +159,43 @@ void Database::Keys(std::string prefix, std::vector<std::string> *keys, KeyNumSt
...
+    if (++slot_id >= HASH_SLOTS_SIZE) break;
+
+    ComposeNamespaceKey(namespace_, "", &ns_prefix);
+    PutFixed16(&ns_prefix, slot_id);
+    ns_prefix.append(prefix);
   }
+
```
Comment (git-hulk) [inline src/redis_db.cc:198]: It Seems uses do...while sentence would be more clear, like below:
```
do {
} while(storage_->IsClusterEnabled() && ++slot_id < HASH_SLOTS_SIZE)
```

**Review by git-hulk — COMMENTED** [review]
cool, overall was good!

File: src/redis_db.cc (hunk 9 above)
Comment (shangxiaoxiong) [reply src/redis_db.cc]: Seems a bug. Should be ns_prefix.

File: src/redis_db.cc (hunk 10 above)
Comment (shangxiaoxiong) [reply src/redis_db.cc:191]: when prefix was empty, keys * is executed. 
Internal cycle `for {} ` can traverse all the keys in a namespace.

**Comment by shangxiaoxiong** [comment]
Comparing with multi-db of Redis,  I prefer that multi-namespace is allowed whether slot-id-encoded is enabled or not but not allowed when cluster mode is enabled.

File: tests/tcl/tests/unit/type/sint.tcl [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+        assert_equal 7 [r SICARD mysi]
+        assert_equal {245 123 5} [r SIREVRANGE mysi 0 3]
+        assert_equal {4 3 2} [r SIREVRANGE mysi 0 3 cursor 5]
+        assert_equal {245} [r SIRANGE mysi 0 3 cursor 123]
+        assert_equal {1 2 3 4} [r SIRANGEBYVALUE mysi 1 (5]
+        assert_equal {5 4 3 2} [r SIREVRANGEBYVALUE mysi 5 (1]
+        assert_equal {1 0 1} [r SIEXISTS mysi 1 88 2]
+        assert_equal {1} [r SIREM mysi 2]
+    }
+}
```
Comment (ShooterIT) [inline tests/tcl/tests/unit/type/sint.tcl]:
```suggestion
}

```

**Comment by ShooterIT** [comment]
I think maybe w can add a special column family named `server`, and special information about server, such as `slot-id-encoded`, `cluster-enabled`, in the future, cluster topology also could be persistent in this CF.

File: src/redis_db.cc (hunk 11 above)
Comment (shangxiaoxiong) [reply src/redis_db.cc:198]: it is not easy to use this way, because we may need to init some variables.

File: src/redis_db.cc (hunk 8 above)
Comment (shangxiaoxiong) [reply src/redis_db.cc]: `ExtractNamespaceKey` also has the same problem. I won't deal with it this time beacuse tuning will affect kvrocks2redis tool. Kvrocks2redis tool does not support slot-id-encoded currently and l will make it support next time and deal with the problem incidentally.

File: src/redis_zset.cc [hunk 13]
```diff
@@ -31,7 +31,8 @@ rocksdb::Status ZSet::Add(const Slice &user_key, uint8_t flags, std::vector<Memb
   std::string member_key;
   std::set<std::string> added_member_keys;
   for (int i = static_cast<int>(mscores->size()-1); i >= 0; i--) {
-    InternalKey(ns_key, (*mscores)[i].member, metadata.version).Encode(&member_key);
+    InternalKey(ns_key, (*mscores)[i].member, metadata.version,
+                storage_->IsSlotIdEncoded()).Encode(&member_key)).Encode(&member_key);
```
Comment (ShooterIT) [inline src/redis_zset.cc]:
```suggestion
                storage_->IsSlotIdEncoded()).Encode(&member_key);
```

File: src/storage.cc [hunk 14]
```diff
@@ -199,7 +204,16 @@ Status Storage::Open(bool read_only) {
   std::vector<std::string> old_column_families;
   auto s = rocksdb::DB::ListColumnFamilies(options, config_->db_dir, &old_column_families);
   if (!s.ok()) return Status(Status::NotOK, s.ToString());
-
+  if (!config_->slot_id_encoded && column_families.size() != old_column_families.size()) {
```
Comment (ShooterIT) [inline src/storage.cc]: please revert this mechanism

File: tests/tcl/tests/unit/scan.tcl [hunk 15]
```diff
@@ -207,4 +207,63 @@ start_server {tags {"scan network"}} {
...
+                set cur [lindex $res 0]
+                set k [lindex $res 1]
+                lappend keys {*}$k
+                if {$cur == 0} break
+            }
+            assert_equal 1000 [llength $keys]
+
+            incr index
+        }
+    } {}
```
Comment (ShooterIT) [inline tests/tcl/tests/unit/scan.tcl]: `{}` is necessary?

File: tests/tcl/tests/unit/keyspace.tcl [hunk 16]
```diff
@@ -107,4 +107,39 @@ start_server {tags {"keyspace"}} {
...
+        # Check KEYS and KEYS MATCH in different namespace
+        set index 0
+        foreach token $tokens {
+            r auth $token
+            assert_equal [lsort [r keys *]]  [lindex $ns_keyspaces $index]
+            assert_equal [lsort [r keys foo*]]  [lindex $foo_prefix_keyspaces $index]
+
+            incr index
+        }
+    } {}
```
Comment (ShooterIT) [inline tests/tcl/tests/unit/keyspace.tcl]: {} is necessary?

**Review by ShooterIT — DISMISSED** [review]

**Comment by ShooterIT** [comment]
@shangxiaoxiong The remaining work
- kvrocks2redis support `slot-id-encoded`, and you also should make sure all keys belonging to one slot are stored together.
- `slot-id-encoded` information are stored in a special CF after we support cluster mode

**Comment by shangxiaoxiong** [comment]
> @shangxiaoxiong The remaining work
> 1、kvrocks2redis support `slot-id-encoded`, and you also should make sure all keys belonging to one slot are stored together.
> 2、`slot-id-encoded` information are stored in a special CF after we support cluster mode

**Review by ShooterIT — DISMISSED** [review]

**Comment by shangxiaoxiong** [comment]
> @shangxiaoxiong The remaining work
> 
> * kvrocks2redis support `slot-id-encoded`, and you also should make sure all keys belonging to one slot are stored together.
> * `slot-id-encoded` information are stored in a special CF after we support cluster mode

Strictly speaking, make sure all keys belonging to one slot of **a namespace** are stored together.

File: kvrocks.conf [hunk 17]
```diff
@@ -200,6 +200,21 @@ max-backup-to-keep 1
...
+#
+# PLEASE NOTE:
+# If you just start to use kvrocks firstly, please don't change it optionally,
+# because cluster mode is dependent with enabling that.
+#
+# If you want to upgrade your old kvrocks version(without encoding slot it into key)
+# whatever by master-replicas replication or restarting using new binary, you must
+# disable this option, otherwise, kvrocks will make data corrupt.
+#
+# Defalut: yes
```
Comment (git-hulk) [inline kvrocks.conf]:
```suggestion
# Default: yes
```

**Review by git-hulk — DISMISSED** [review]

**Review by ShooterIT — APPROVED** [review]

**Review by git-hulk — APPROVED** [review]

**Review by karelrooted — APPROVED** [review]

**Comment by ShooterIT** [comment]
@shangxiaoxiong Finally, merged, another thing i find is that we should tell users our cursor of scan is not a number instead of a string that has a special prefix `-`.

**Pull request merged**